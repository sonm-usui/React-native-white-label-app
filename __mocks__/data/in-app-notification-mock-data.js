// @flow

import { MESSAGE_TYPE } from '../../src/api/api-constants'

export const inAppNotificationMockData = {
  notification: {
    senderName: 'Anonymous',
    senderImage: 'https://localhost/image.png',
    senderDID: 'senderDID1',
    identifier: 'receiverPairwiseDID1',
    title: 'You received a PROOF REQUEST',
    text: 'Test wants you to share Address',
    messageType: MESSAGE_TYPE.PROOF_REQUEST,
    messageId: 'messageId1',
  },
}

export const inAppNotificationEmptyMockData = {
  notification: null,
}
