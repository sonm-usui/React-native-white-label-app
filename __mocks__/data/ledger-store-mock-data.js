// @flow

import moment from 'moment'
import { STORE_STATUS } from '../../src/common/type-common'

export const transferFees = {
  transfer: '0.00001',
  refreshTime: moment().format(),
}

export const ledgerStoreWithTransferFees = {
  fees: {
    data: transferFees,
    status: STORE_STATUS.SUCCESS,
    error: null,
  },
}
