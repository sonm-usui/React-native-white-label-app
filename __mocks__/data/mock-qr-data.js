// @flow

export const mockAriesV1InvitationUrl =
  'http://localhost:80/?c_i=somebase64encodedstring'

export const mockAriesV1QrCode = {
  '@id': '7bf4f041-5ed4-4246-b309-396aeae830ab',
  '@type': 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/connections/1.0/invitation',
  label: 'alice',
  recipientKeys: ['6B5wQFpWbwJYB7qQYWydSwvCNKGoTZq6KvUe8TWBgPFY'],
  routingKeys: [
    '7re7C6zmg3p1r1hzHZSQJgt27wbycLeuzgopq1xNZqs6',
    '7G3LhXFKXKTMv7XGx1Qc9wqkMbwcU2iLBHL8x1JXWWC2',
  ],
  serviceEndpoint: 'https://agency.com/agency/msg',
}

export const mockAriesV1QrCodeNoRecipientKeys = {
  '@id': '7bf4f041-5ed4-4246-b309-396aeae830ab',
  '@type': 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/connections/1.0/invitation',
  label: 'alice',
  routingKeys: [
    '7re7C6zmg3p1r1hzHZSQJgt27wbycLeuzgopq1xNZqs6',
    '7G3LhXFKXKTMv7XGx1Qc9wqkMbwcU2iLBHL8x1JXWWC2',
  ],
  serviceEndpoint: 'https://agency.com/agency/msg',
}

export const mockAriesV1QrCodeNoRoutingKeys = {
  '@id': '7bf4f041-5ed4-4246-b309-396aeae830ab',
  '@type': 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/connections/1.0/invitation',
  label: 'alice',
  recipientKeys: ['6B5wQFpWbwJYB7qQYWydSwvCNKGoTZq6KvUe8TWBgPFY'],
  serviceEndpoint: 'https://agency.com/agency/msg',
}

export const mockAriesV1QrCodeNullRoutingKeys = {
  '@id': '7bf4f041-5ed4-4246-b309-396aeae830ab',
  '@type': 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/connections/1.0/invitation',
  label: 'alice',
  recipientKeys: ['6B5wQFpWbwJYB7qQYWydSwvCNKGoTZq6KvUe8TWBgPFY'],
  routingKeys: null,
  serviceEndpoint: 'https://agency.com/agency/msg',
}

export const mockAriesV1QrCodeNoLabel = {
  '@id': '7bf4f041-5ed4-4246-b309-396aeae830ab',
  '@type': 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/connections/1.0/invitation',
  recipientKeys: ['6B5wQFpWbwJYB7qQYWydSwvCNKGoTZq6KvUe8TWBgPFY'],
  routingKeys: null,
  serviceEndpoint: 'https://agency.com/agency/msg',
}

export const mockAriesV1QrCodeNullLabel = {
  '@id': '7bf4f041-5ed4-4246-b309-396aeae830ab',
  '@type': 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/connections/1.0/invitation',
  label: null,
  recipientKeys: ['6B5wQFpWbwJYB7qQYWydSwvCNKGoTZq6KvUe8TWBgPFY'],
  routingKeys: null,
  serviceEndpoint: 'https://agency.com/agency/msg',
}

export const mockEphemeralProofRequestQrCode = {
  '@id': '7bf4f041-5ed4-4246-b309-396aeae830bc',
  '@type': 'did:sov:BzCbsNYhMrjHiqZDTUASHg;spec/connections/1.0/invitation',
  'request_presentations~attach': [
    {
      '@id': '7bf4f041-5ed4-4246-b309-396aeae830cd',
      'mime-type': 'application/json',
      data: {
        base64: 'oihkadnkfnakfadfda=/',
      },
    },
  ],
  comment: 'Sender Name',
  '~service': {
    recipientKeys: ['6B5wQFpWbwJYB7qQYWydSwvCNKGoTZq6KvUe8TWBgPFY'],
    routingKeys: [
      '7re7C6zmg3p1r1hzHZSQJgt27wbycLeuzgopq1xNZqs6',
      '7G3LhXFKXKTMv7XGx1Qc9wqkMbwcU2iLBHL8x1JXWWC2',
    ],
    serviceEndpoint: 'https://agency.com/agency/msg',
  },
}
