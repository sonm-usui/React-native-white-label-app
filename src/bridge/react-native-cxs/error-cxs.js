// @flow
export const CONNECTION_ALREADY_EXISTS = '1062'
export const CREDENTIAL_DEFINITION_NOT_FOUND = '1036'
export const SCHEMA_NOT_FOUND = '1040'
export const INVALID_CREDENTIAL_OFFER = '1043'
export const ACTION_IS_NOT_SUPPORTED = '1103'
export const CLOUD_AGENT_UNAVAILABLE = '1010'
export const NO_POOL_OPEN = '1030'
export const CREDENTIAL_SCHEMA_NOT_FOUND = '1031'
export const WALLET_ITEM_NOT_FOUND = '1073'
export const INVALID_AGENCY_RESPONSE = '1020'
export const NOT_READY = '1005'
export const CONNECTION_UPGRADE_NOT_NEEDED = '1065'

export const CONNECTION_ALREADY_EXISTS_MESSAGE =
  'The connection invitation has already been accepted. ' +
  'Please, request a new invitation.'

export const CREDENTIAL_DEFINITION_NOT_FOUND_MESSAGE =
  'Definition for offered Credential not found on the network application is connected to.'

export const SCHEMA_NOT_FOUND_MESSAGE =
  'Schema for offered Credential not found on the network application is connected to.'

export const INVALID_CREDENTIAL_OFFER_MESSAGE =
  'Offered Credential does not match to its schema definition.'
