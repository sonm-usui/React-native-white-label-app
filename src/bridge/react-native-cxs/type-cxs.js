// @flow

import type { UserOneTimeInfo } from '../../store/user/type-user-store'
import type { AgencyPoolConfig } from '../../store/type-config-store'
import type { ClaimOfferPushPayload } from '../../push-notification/type-push-notification'
import type { ProprietaryConnectionInvitation } from '../../invitation/type-invitation'

export type AriesProtocolConfig = {
  protocol_type: string,
}

export type VcxProvision = {
  agency_url: string,
  agency_did: string,
  agency_verkey: string,
} & AriesProtocolConfig

export type VcxProvisionResult = {
  wallet_name: string,
  wallet_key: string,
  agency_endpoint: string,
  agency_did: string,
  agency_verkey: string,
  // myOneTimeDid
  sdk_to_remote_did: string,
  // myOneTimeVerificationKey
  sdk_to_remote_verkey: string,
  // oneTimeAgencyDid
  institution_did: string,
  // oneTimeAgencyVerificationKey
  institution_verkey: string,
  // myOneTimeAgentDid
  remote_to_sdk_did: string,
  // myOneTimeAgentVerificationKey
  remote_to_sdk_verkey: string,
}

export type CxsInitConfig = UserOneTimeInfo & AgencyPoolConfig

export type VcxInitConfig = {
  agency_endpoint: string,
  agency_did: string,
  agency_verkey: string,
  wallet_key: string,
  wallet_name: string,
  remote_to_sdk_did: string,
  remote_to_sdk_verkey: string,
  sdk_to_remote_did: string,
  sdk_to_remote_verkey: string,
  institution_did: string,
  institution_verkey: string,
} & AriesProtocolConfig

export type VcxPushTokenConfig = {
  id: string,
  value: string,
}

export type CxsPushTokenConfig = {
  uniqueId: string,
  pushToken: string,
}

export type CxsPoolConfig = {
  poolConfig: string,
}

export type CxsPoolConfigWithGenesisPath = CxsPoolConfig & {
  genesis_path: string,
}

export type VcxPoolInitConfig = {
  genesis_path: string,
  pool_name?: string,
}

export type VcxCreateConnection = {
  source_id: string,
  invite_details: ProprietaryConnectionInvitation,
}

export type VcxConnectionConnectResult = {|
  source_id: string,
  pw_did: string,
  pw_verkey: string,
  uuid: string,
  endpoint: string,
  agent_did: string,
  agent_vk: string,
  their_pw_did: string,
  their_pw_verkey: string,
|}

export type VcxCredentialOffer = {
  msg_type: string,
  version: string,
  to_did: string,
  from_did: string,
  libindy_offer: string,
  cred_def_id: string,
  credential_attrs: { [string]: Array<string> },
  schema_seq_no: number,
  claim_name: string,
  claim_id: string,
  payment_address?: ?string,
  price?: ?string,
}

export type CxsCredentialOfferResult = {
  claimHandle: number,
  claimOffer: ClaimOfferPushPayload,
}

export type VcxClaimInfo = {
  credential_id?: string,
  credential?: string,
  credential_offer?: string,
}

export type WalletPoolName = {
  walletName: string,
  poolName: string,
}

export type UTXO = {
  amount: number,
  extra: string,
  paymentAddress: string,
  source: string,
}

export type PaymentAddress = {
  address: string,
  balance: number,
  utxo: UTXO[],
}

export type WalletTokenInfo = {
  balance_str: string,
  addresses: PaymentAddress[],
}

export const smallDeviceMemory = 1073741824

export type SignDataResponse = {
  data: string,
  signature: string,
}

export const signDataResponseSchema = {
  type: 'object',
  properties: {
    data: { type: 'string', minLength: 1 },
    signature: { type: 'string', minLength: 1 },
  },
  required: ['data', 'signature'],
}
