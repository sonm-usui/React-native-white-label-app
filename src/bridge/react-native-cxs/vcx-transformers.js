// @flow

import type { GenericObject } from '../../common/type-common'
import type { AgencyPoolConfig } from '../../store/type-config-store'
import type {
  VcxProvision,
  VcxProvisionResult,
  CxsInitConfig,
  VcxInitConfig,
  CxsPushTokenConfig,
  VcxPushTokenConfig,
  VcxCreateConnection,
  VcxConnectionConnectResult,
  VcxCredentialOffer,
  WalletPoolName,
} from './type-cxs'
import type { UserOneTimeInfo } from '../../store/user/type-user-store'
import type { InvitationPayload } from '../../invitation/type-invitation'
import type { CredentialOffer } from '../../claim-offer/type-claim-offer'
import type { MyPairwiseInfo } from '../../store/type-connection-store'
import type { ClaimOfferPushPayload } from '../../push-notification/type-push-notification'
import { getWalletKey } from '../../services/storage'
import { appName, vcxPushType } from '../../external-imports'
import DeviceInfo from 'react-native-device-info'
import { flattenAsync } from '../../common/flatten-async'
import { Platform } from 'react-native'
import { getDeviceAttestation } from '../../start-up/device-check-saga'
import { flatJsonParse } from '../../common/flat-json-parse'
import { toUtf8FromBase64 } from './RNCxs'
import type {
  AdditionalProofDataPayload,
  AriesPresentationRequest,
} from '../../proof-request/type-proof-request'
import { convertProofRequestPushPayloadToAppProofRequest } from '../../push-notification/push-notification-store'

export const paymentHandle = 0
const commonConfigParams = {
  protocol_type: '4.0',
}

export async function convertAgencyConfigToVcxProvision(
  config: AgencyPoolConfig,
  walletPoolName: WalletPoolName
): Promise<VcxProvision> {
  const wallet_key = await getWalletKey()

  return {
    agency_url: config.agencyUrl,
    agency_did: config.agencyDID,
    agency_verkey: config.agencyVerificationKey,
    wallet_name: walletPoolName.walletName,
    wallet_key,
    agent_seed: null,
    enterprise_seed: null,
    ...commonConfigParams,
  }
}

export async function addAttestation(token: string) {
  const [parseError, parsedToken] = flatJsonParse(token)
  if (parseError || !parsedToken || !parsedToken.nonce) {
    return token
  }

  const [attestationError, attestationSignature] = await flattenAsync(
    getDeviceAttestation
  )(parsedToken.nonce)
  if (attestationError || !attestationSignature) {
    return token
  }

  return JSON.stringify({
    ...parsedToken,
    attestationAlgorithm: Platform.OS === 'ios' ? 'DeviceCheck' : 'SafetyNet',
    attestationData: attestationSignature,
  })
}

export function convertVcxProvisionResultToUserOneTimeInfo(
  provision: VcxProvisionResult
): UserOneTimeInfo {
  return {
    oneTimeAgencyDid: provision.institution_did,
    oneTimeAgencyVerificationKey: provision.institution_verkey,
    myOneTimeDid: provision.sdk_to_remote_did,
    myOneTimeVerificationKey: provision.sdk_to_remote_verkey,
    myOneTimeAgentDid: provision.remote_to_sdk_did,
    myOneTimeAgentVerificationKey: provision.remote_to_sdk_verkey,
  }
}

export async function convertCxsInitToVcxInit(
  init: CxsInitConfig,
  walletPoolName: WalletPoolName
): Promise<VcxInitConfig> {
  const wallet_key = await getWalletKey()
  const [, deviceName] = await flattenAsync(DeviceInfo.getDeviceName)()

  return {
    agency_endpoint: init.agencyUrl,
    agency_did: init.agencyDID,
    agency_verkey: init.agencyVerificationKey,
    wallet_name: walletPoolName.walletName,
    wallet_key,
    remote_to_sdk_did: init.myOneTimeAgentDid,
    remote_to_sdk_verkey: init.myOneTimeAgentVerificationKey,
    sdk_to_remote_did: init.myOneTimeDid,
    sdk_to_remote_verkey: init.myOneTimeVerificationKey,
    // TODO: These should be removed after we sdk team fix these as optional
    institution_name: deviceName || appName,
    institution_logo_url: 'https://try.connect.me/img/CMicon@3x.png',
    institution_did: init.oneTimeAgencyDid,
    institution_verkey: init.oneTimeAgencyVerificationKey,
    ...commonConfigParams,
  }
}

export function convertCxsPushConfigToVcxPushTokenConfig(
  pushConfig: CxsPushTokenConfig
): VcxPushTokenConfig {
  return {
    type: vcxPushType,
    id: pushConfig.uniqueId,
    value: pushConfig.pushToken,
  }
}

export function convertInvitationToVcxConnectionCreate(
  invitation: InvitationPayload
): VcxCreateConnection {
  return {
    source_id: invitation.requestId,
    invite_details: {
      connReqId: invitation.requestId,
      // TODO: Add status code to be available in invitation payload
      // for now, it would always be MS-102
      statusCode: 'MS-102',
      senderDetail: invitation.senderDetail,
      senderAgencyDetail: invitation.senderAgencyDetail,
      targetName: invitation.senderName,
      // hard coding this for now, because this field does not matter anywhere for processing
      // and it will always be message sent for the purpose of connection create
      statusMsg: 'message sent',
      ...(invitation.version ? { version: invitation.version } : {}),
    },
  }
}

export function convertVcxConnectionToAppConnection(
  vcxConnection: VcxConnectionConnectResult
): MyPairwiseInfo {
  return {
    myPairwiseDid: vcxConnection.pw_did,
    myPairwiseVerKey: vcxConnection.pw_verkey,
    myPairwiseAgentDid: vcxConnection.agent_did,
    myPairwiseAgentVerKey: vcxConnection.agent_vk,
    myPairwisePeerVerKey: vcxConnection.their_pw_verkey,
    senderDID: vcxConnection.their_pw_did,
  }
}

export function convertLegacyClaimOfferToAppClaimOffer(
  vcxCredentialOffer: VcxCredentialOffer
): ClaimOfferPushPayload {
  return {
    msg_type: vcxCredentialOffer.msg_type,
    version: vcxCredentialOffer.version,
    to_did: vcxCredentialOffer.to_did,
    from_did: vcxCredentialOffer.from_did,
    claim: vcxCredentialOffer.credential_attrs,
    claim_name: vcxCredentialOffer.claim_name || 'Credential',
    claim_def_id: extractCredDefIdFromIndyCredentialOffer(
      vcxCredentialOffer.libindy_offer
    ),
    schema_seq_no: vcxCredentialOffer.schema_seq_no,
    issuer_did: vcxCredentialOffer.from_did,
    // should override it when generating claim offer object
    remoteName: '',
    price: vcxCredentialOffer.price,
  }
}

export async function convertAriesCredentialOfferToAppClaimOffer(
  credentialOffer: CredentialOffer
) {
  let claim: GenericObject = {}

  // check whether data is valid base64 string
  const [
    decodedCredentialOfferError,
    decodedCredentialOffer,
  ] = await flattenAsync(toUtf8FromBase64)(
    credentialOffer['offers~attach'][0].data.base64
  )
  if (decodedCredentialOfferError || decodedCredentialOffer === null) {
    return null
  }

  // check whether decoded data is valid json or not
  const [parseCredentialOfferError, parsedCredentialOffer] = flatJsonParse(
    decodedCredentialOffer
  )
  if (parseCredentialOfferError || parsedCredentialOffer === null) {
    return null
  }

  for (const a of credentialOffer.credential_preview.attributes) {
    claim[a.name] = a.value
  }

  return {
    msg_type: 'credential-offer',
    version: '0.1',
    to_did: '',
    from_did: '',
    claim: claim,
    claim_name:
      credentialOffer.comment ||
      extractCredentialNameFromSchemaId(parsedCredentialOffer['schema_id']) ||
      credentialOffer['~alias']?.label ||
      'Credential',
    claim_def_id: parsedCredentialOffer['cred_def_id'],
    schema_seq_no: 0,
    issuer_did: '',
    remoteName:
      credentialOffer.comment ||
      credentialOffer['~alias']?.label ||
      'Unnamed Connection',
  }
}

export async function convertAriesProofRequestToCxsProofRequest(
  presentationRequest: AriesPresentationRequest,
  senderName?: string
): Promise<null | AdditionalProofDataPayload> {
  // we still need to get data from base64
  const [decodeProofRequestError, decodedProofRequest] = await flattenAsync(
    toUtf8FromBase64
  )(presentationRequest['request_presentations~attach'][0].data.base64)
  if (decodeProofRequestError || decodedProofRequest === null) {
    return null
  }

  // check whether decoded data is valid json or not
  const [parseProofRequestError, parsedProofRequest] = flatJsonParse(
    decodedProofRequest
  )
  if (parseProofRequestError || parsedProofRequest === null) {
    return null
  }

  return convertProofRequestPushPayloadToAppProofRequest({
    proof_request_data: {
      ...parsedProofRequest,
    },
    '@topic': { mid: 0, tid: 0 },
    '@type': { name: 'proof_request', version: '0.1' },
    remoteName:
      senderName || presentationRequest.comment || 'Verification Request',
    proofHandle: 0,
  })
}

const extractCredDefIdFromIndyCredentialOffer = (offer: string | null) => {
  try {
    if (!offer) {
      return null
    }
    const parsed = JSON.parse(offer)
    return parsed['cred_def_id']
  } catch (e) {
    return null
  }
}

export const extractCredentialNameFromSchemaId = (
  schemaId: string | null
): string | null => {
  if (!schemaId) {
    return null
  }
  const parts = schemaId.split(':')
  if (parts.length === 4) {
    // regular schema id
    return parts[2]
  }
  if (parts.length === 6) {
    // fully-qualified schema id
    return parts[4]
  }
  if (parts.length === 8) {
    // fully-qualified schema id
    return parts[6]
  }
  return null
}
