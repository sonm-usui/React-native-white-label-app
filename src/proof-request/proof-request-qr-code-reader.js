// @flow

import type {
  AdditionalProofDataPayload,
  AriesPresentationRequest,
  QrCodeEphemeralProofRequest,
} from './type-proof-request'

import { flattenAsync } from '../common/flatten-async'
import { toUtf8FromBase64 } from '../bridge/react-native-cxs/RNCxs'
import { schemaValidator } from '../services/schema-validator'
import { QR_CODE_TYPES } from '../components/qr-scanner/type-qr-scanner'
import { convertProofRequestPushPayloadToAppProofRequest } from '../push-notification/push-notification-store'
import { flatJsonParse } from '../common/flat-json-parse'

export async function validateEphemeralProofQrCode(
  qrCode: Object,
): Promise<[
    null | string,
    null | { type: string, proofRequest: QrCodeEphemeralProofRequest }
]> {
  // MSDK supports three types of data for ephemeral proof request
  // 1. Url that returns an ephemeral proof request
  // 2. base64 encoded ephemeral proof request
  // 3. json formatted ephemeral proof request

  // Url downloading is already handled by qr-scanner
  // which if url is passed in qr code, then qr-scanner downloads the data
  // and passes into this function
  // if qr code is not url, then qr code is simply passed to this function
  // So, we need to check if qr code is either base64 encoded, or json formatted

  // we need to keep track of original message as well, that we can pass to vcx
  let originalMessage = JSON.stringify(qrCode)

  // check whether the response is valid json
  // first try parsing decodedResponse
  const ephemeralProofRequest = qrCode
  // check if json has correct data
  if (
    !schemaValidator.validate(
      ephemeralProofRequestSchema,
      ephemeralProofRequest,
    )
  ) {
    return ['EPR-002::Invalid data.', null]
  }

  // Now, we know that format of ephemeral proof request is valid

  // we still need to get data from base64
  const [decodeProofRequestError, decodedProofRequest] = await flattenAsync(
    toUtf8FromBase64,
  )(ephemeralProofRequest['request_presentations~attach'][0].data.base64)
  if (decodeProofRequestError || decodedProofRequest === null) {
    return ['EPR-003::Invalid proof request.', null]
  }

  // check whether decoded data is valid json or not
  const [parseProofRequestError, parsedProofRequest] = flatJsonParse(
    decodedProofRequest,
  )
  if (parseProofRequestError || parsedProofRequest === null) {
    return ['EPR-004::Invalid proof request format.', null]
  }

  // check if parsed json is valid
  if (!schemaValidator.validate(proofRequestDataSchema, parsedProofRequest)) {
    return ['EPR-005::Invalid proof request data.', null]
  }

  return [
    null,
    {
      type: QR_CODE_TYPES.EPHEMERAL_PROOF_REQUEST_V1,
      proofRequest: {
        ephemeralProofRequest,
        proofRequestPayload: convertProofRequestPushPayloadToAppProofRequest({
          proof_request_data: {
            ...parsedProofRequest,
          },
          '@topic': { mid: 0, tid: 0 },
          '@type': { name: 'proof_request', version: '0.1' },
          remoteName: ephemeralProofRequest.comment || 'Unnamed Connection',
          proofHandle: 0,
          ephemeralProofRequest: originalMessage,
        }),
      },
    },
  ]
}

export async function validateOutofbandProofRequestQrCode(
  presentationRequest: AriesPresentationRequest,
  senderName?: string
): Promise<[null | string, null | AdditionalProofDataPayload]> {
  // we still need to get data from base64
  const [decodeProofRequestError, decodedProofRequest] = await flattenAsync(
    toUtf8FromBase64
  )(presentationRequest['request_presentations~attach'][0].data.base64)
  if (decodeProofRequestError || decodedProofRequest === null) {
    return ['OOBPR-003::Invalid proof request.', null]
  }

  // check whether decoded data is valid json or not
  const [parseProofRequestError, parsedProofRequest] = flatJsonParse(
    decodedProofRequest
  )
  if (parseProofRequestError || parsedProofRequest === null) {
    return ['OOBPR-004::Invalid proof request format.', null]
  }

  // check if parsed json is valid

  if (!schemaValidator.validate(proofRequestDataSchema, parsedProofRequest)) {
    return ['OOBPR-005::Invalid proof request data.', null]
  }

  return [
    null,
    convertProofRequestPushPayloadToAppProofRequest({
      proof_request_data: {
        ...parsedProofRequest,
      },
      '@topic': { mid: 0, tid: 0 },
      '@type': { name: 'proof_request', version: '0.1' },
      remoteName: senderName || presentationRequest.comment || 'Verification Request',
      proofHandle: 0,
      outofbandProofRequest: JSON.stringify(presentationRequest),
    }),
  ]
}

export const ephemeralProofRequestSchema = {
  type: 'object',
  properties: {
    '@id': { type: 'string', minLength: 4, maxLength: 1024 },
    '@type': { type: 'string', minLength: 3, maxLength: 500 },
    'request_presentations~attach': {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          '@id': { type: 'string', minLength: 4, maxLength: 1024 },
          'mime-type': {
            type: 'string',
            minLength: 3,
            maxLength: 100,
            enum: ['application/json'],
          },
          data: {
            type: 'object',
            properties: {
              base64: { type: 'string', minLength: 10 },
            },
            required: ['base64'],
          },
        },
        required: ['@id', 'data'],
      },
      minLength: 1,
    },
    comment: { type: ['null', 'string'] },
    '~service': {
      type: 'object',
      properties: {
        recipientKeys: {
          type: 'array',
          items: {
            type: 'string',
          },
          minLength: 1,
        },
        routingKeys: {
          type: ['null', 'array'],
          items: { type: 'string' },
        },
        serviceEndpoint: { type: 'string', minLength: 10 },
      },
      required: ['recipientKeys', 'serviceEndpoint'],
    },
  },
  required: ['request_presentations~attach', '~service', '@id', '@type'],
}

const proofRequestDataSchema = {
  type: 'object',
  properties: {
    name: { type: 'string', minLength: 1, maxLength: 300 },
    version: { type: 'string', minLength: 1, maxLength: 200 },
    requested_attributes: {
      type: 'object',
      patternProperties: {
        '.*': {
          type: 'object',
          properties: {
            name: { type: ['null', 'string'] },
            names: { type: ['null', 'array'] },
            restrictions: { type: ['array', 'object'] },
          },
        },
      },
    },
  },
  required: ['name', 'version', 'requested_attributes'],
}

export const presentationProposalSchema = {
  type: 'object',
  properties: {
    '@id': { type: 'string', minLength: 4, maxLength: 1024 },
    '@type': { type: 'string', minLength: 3, maxLength: 500 },
    comment: { type: ['null', 'string'] },
    presentation_proposal: {
      type: 'object',
      '@type': { type: 'string', minLength: 3, maxLength: 500 },
      attributes: {
        type: 'array',
        items: {
          type: 'object',
          properties: {
            name: { type: 'string', minLength: 1, maxLength: 1024 },
            value: { type: ['null', 'string'] },
            cred_def_id: { type: ['null', 'string'] },
          },
          required: ['name'],
        },
      },
      predicates: {
        type: 'array',
        items: {
          type: 'object',
          properties: {
            name: { type: 'string', minLength: 1, maxLength: 1024 },
            predicate: { type: ['null', 'string'] },
            threshold: { type: ['null', 'string'] },
            cred_def_id: { type: ['null', 'string'] },
          },
          required: ['name'],
        },
      },
    },
  },
  required: ['@id', '@type', 'presentation_proposal'],
}
