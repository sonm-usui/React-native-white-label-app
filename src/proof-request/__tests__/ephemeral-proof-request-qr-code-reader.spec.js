// @flow

import 'react-native'

import * as vcx from '../../bridge/react-native-cxs/RNCxs'
import { validateEphemeralProofQrCode } from '../proof-request-qr-code-reader'
import { mockEphemeralProofRequestQrCode } from '../../../__mocks__/data/mock-qr-data'
import { originalProofRequestData } from '../../../__mocks__/static-data'

describe('ephemeral-proof-request-qr-code-reader', () => {
  it('should return error code EPR-002', async () => {
    const decodeSpy = jest.spyOn(vcx, 'toUtf8FromBase64')
    decodeSpy.mockImplementation(() =>
      Promise.resolve(
        JSON.stringify({ ...mockEphemeralProofRequestQrCode, '@id': 1 })
      )
    )

    const [error] = await validateEphemeralProofQrCode(
      '{"key":"this json should not be used, and decode response should be used"}'
    )

    expect(error).toBe('EPR-002::Invalid data.')

    decodeSpy.mockReset()
    decodeSpy.mockRestore()
  })

  it('should return error code EPR-003', async () => {
    const decodeSpy = jest.spyOn(vcx, 'toUtf8FromBase64')
    decodeSpy
      .mockImplementationOnce(() => Promise.reject('invalid base64 qr code'))
      .mockImplementationOnce(() =>
        Promise.reject('some error for decoding proof request data')
      )

    const [error] = await validateEphemeralProofQrCode(
      JSON.stringify(mockEphemeralProofRequestQrCode)
    )

    expect(error).toBe('EPR-002::Invalid data.')

    decodeSpy.mockReset()
    decodeSpy.mockRestore()
  })

  it('should return error code EPR-004', async () => {
    const decodeSpy = jest.spyOn(vcx, 'toUtf8FromBase64')
    decodeSpy
      .mockImplementationOnce(() => Promise.reject('invalid base64 qr code'))
      .mockImplementationOnce(() =>
        Promise.resolve('invalid data which is not json format')
      )

    const [error] = await validateEphemeralProofQrCode(
      mockEphemeralProofRequestQrCode
    )

    expect(error).toBe('EPR-003::Invalid proof request.')

    decodeSpy.mockReset()
    decodeSpy.mockRestore()
  })

  it('should return error code EPR-005', async () => {
    const decodeSpy = jest.spyOn(vcx, 'toUtf8FromBase64')
    decodeSpy
      .mockImplementationOnce(() => Promise.reject('invalid base64 qr code'))
      .mockImplementationOnce(() =>
        Promise.resolve(
         {
            ...originalProofRequestData,
            requested_attributes: {
              name: {
                name: 'address',
                restrictions: 'some restrictions',
              },
            },
          }
        )
      )

    const [error] = await validateEphemeralProofQrCode(
      mockEphemeralProofRequestQrCode
    )

    expect(error).toBe('EPR-003::Invalid proof request.')

    decodeSpy.mockReset()
    decodeSpy.mockRestore()
  })

  it('should match valid ephemeral proof request qr code', async () => {
    const decodeSpy = jest.spyOn(vcx, 'toUtf8FromBase64')
    decodeSpy
      .mockImplementationOnce(() => Promise.reject('invalid base64 qr code'))
      .mockImplementationOnce(() =>
        Promise.resolve(
          JSON.stringify({
            ...originalProofRequestData,
          })
        )
      )

    const [, ephemeralProofRequest] = await validateEphemeralProofQrCode(
      JSON.stringify(mockEphemeralProofRequestQrCode)
    )

    expect(ephemeralProofRequest).toMatchSnapshot()

    decodeSpy.mockReset()
    decodeSpy.mockRestore()
  })

  it('should match valid ephemeral proof request with null comment, and null routingKeys', async () => {
    const decodeSpy = jest.spyOn(vcx, 'toUtf8FromBase64')
    decodeSpy
      .mockImplementationOnce(() => Promise.reject('invalid base64 qr code'))
      .mockImplementationOnce(() =>
        Promise.resolve(
          JSON.stringify({
            ...originalProofRequestData,
          })
        )
      )

    const [, ephemeralProofRequest] = await validateEphemeralProofQrCode(
      JSON.stringify({
        ...mockEphemeralProofRequestQrCode,
        comment: null,
        '~service': {
          ...mockEphemeralProofRequestQrCode['~service'],
          routingKeys: null,
        },
      })
    )

    expect(ephemeralProofRequest).toMatchSnapshot()

    decodeSpy.mockReset()
    decodeSpy.mockRestore()
  })
})
