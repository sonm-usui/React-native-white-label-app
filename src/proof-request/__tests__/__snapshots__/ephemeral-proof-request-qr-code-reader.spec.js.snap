// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ephemeral-proof-request-qr-code-reader should match valid ephemeral proof request qr code 1`] = `null`;

exports[`ephemeral-proof-request-qr-code-reader should match valid ephemeral proof request with null comment, and null routingKeys 1`] = `null`;
